### PR TITLE
IGNITE-22182 : Move snapshop metadata validation to SnapshotMetadataVerificationTask

### PR DESCRIPTION
**Pull request by Vladsz83**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java [hunk 1]
```diff
@@ -88,25 +96,74 @@ public MetadataVerificationJob(SnapshotMetadataVerificationTaskArg arg) {
...
             IgniteSnapshotManager snpMgr = ignite.context().cache().context().snapshotMgr();
 
-            List<SnapshotMetadata> snpMeta = snpMgr.readSnapshotMetadatas(arg.snapshotName(), arg.snapshotPath());
+            List<SnapshotMetadata> metas = snpMgr.readSnapshotMetadatas(arg.snapshotName(), arg.snapshotPath());
+
+            for (SnapshotMetadata meta : metas)
+                checkMeta(meta, ignite.context().config().getEncryptionSpi().masterKeyDigest());
 
             if (arg.incrementIndex() > 0) {
-                List<SnapshotMetadata> metas = snpMeta.stream()
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java:97]: Let's revert the name of the variable. You actually start using `metas` name for another variable. It's a little bit misleading.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java [hunk 2]
```diff
@@ -88,25 +96,74 @@ public MetadataVerificationJob(SnapshotMetadataVerificationTaskArg arg) {
...
+                List<SnapshotMetadata> locNodeMetas = metas.stream()
                     .filter(m -> m.consistentId().equals(ignite.localNode().consistentId()))
                     .collect(Collectors.toList());
 
-                if (metas.size() != 1) {
-                    throw new IgniteException("Failed to find snapshot metafile [metas=" + metas +
-                        ", snpName=" + arg.snapshotName() + ", snpPath=" + arg.snapshotPath() + ']');
+                if (locNodeMetas.size() != 1) {
+                    throw new IgniteException("Failed to find single snapshot metafile for local node [locNodeId="
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java:113]: Let's write here that incremental snapshots requires single meta file per node, because incremental snapshots don't support restoring on different topology.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTaskArg.java [hunk 3]
```diff
@@ -76,18 +81,27 @@ public int incrementIndex() {
...
+    public Collection<Integer> grpIds() {
+        return grpIds;
+    }
+
     /** {@inheritDoc} */
     @Override protected void writeExternalData(ObjectOutput out) throws IOException {
         U.writeString(out, snpName);
         U.writeString(out, snpPath);
         out.writeInt(incIdx);
+        U.writeCollection(out, grpIds);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTaskArg.java:96]: Should change protocol version to V2.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java [hunk 4]
```diff
@@ -402,12 +402,8 @@ public IgniteFutureImpl<Void> start(
                 return;
             }
 
-            if (!reqGrpIds.isEmpty()) {
-                finishProcess(fut0.rqId, new IllegalArgumentException(OP_REJECT_MSG + "Cache group(s) was not " +
-                    "found in the snapshot [groups=" + reqGrpIds.values() + ", snapshot=" + snpName + ']'));
-
-                return;
-            }
+            assert reqGrpIds.isEmpty() : "Cache group(s) was not found in the snapshot [groups=" + reqGrpIds.values()
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java:405]: Why this change from exception to assert?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java [hunk 5]
```diff
@@ -88,25 +96,74 @@ public MetadataVerificationJob(SnapshotMetadataVerificationTaskArg arg) {
...
+            if (snpMasterKeyDigest != null && !Arrays.equals(snpMasterKeyDigest, masterKeyDigest)) {
+                throw new IllegalStateException("Snapshot '" + meta.snapshotName() + "' has different master " +
+                    "key digest. To restore this snapshot, start Ignite with the same master key.");
+            }
+
+            Collection<Integer> grpIds = new HashSet<>(F.isEmpty(arg.grpIds()) ? meta.cacheGroupIds() : arg.grpIds());
+
+            if (meta.hasCompressedGroups() && grpIds.stream().anyMatch(meta::isGroupWithCompression)) {
+                try {
+                    if (ignite.context().compress() == null)
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java:143]: Do actually we need this check? I suppose that user should add ignite-compress module to classpath and this message is in the catch exception block below. Let's catch the NPE with the IgniteCheckedException together.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java (hunk 4 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java:405]: > Why this change from exception to assert?

Because this duplicates the check within MetadataVerificationJob#checkMeta(SnapshotMetadata meta) and must not occure here.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java [hunk 6]
```diff
@@ -187,39 +249,62 @@ private void checkWalSegments(IncrementalSnapshotMetadata meta, long startWalSeg
     }
 
     /** {@inheritDoc} */
-    @Override public @Nullable SnapshotMetadataVerificationTaskResult reduce(List<ComputeJobResult> results) throws IgniteException {
+    @Override public SnapshotMetadataVerificationTaskResult reduce(
+        List<ComputeJobResult> results) throws IgniteException {
         Map<ClusterNode, List<SnapshotMetadata>> reduceRes = new HashMap<>();
-        Map<ClusterNode, Exception> exs = new HashMap<>();
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java:192]: Useless renaming

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java [hunk 7]
```diff
@@ -187,39 +249,62 @@ private void checkWalSegments(IncrementalSnapshotMetadata meta, long startWalSeg
...
                 if (!first.sameSnapshot(meta)) {
-                    exs.put(res.getNode(),
+                    exceptions.put(res.getNode(),
                         new IgniteException("An error occurred during comparing snapshot metadata from cluster nodes " +
                             "[first=" + first + ", meta=" + meta + ", nodeId=" + res.getNode().id() + ']'));
 
                     continue;
                 }
 
-                reduceRes.computeIfAbsent(res.getNode(), n -> new ArrayList<>())
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java:217]: useless change

**Review by timoninmaxim — APPROVED** [review]
